wallet-core/src/ffi.rs

```
@@ -196,10 +263,215 @@ pub unsafe fn pick_notes(
+
+    tx.set_proof(proof.to_vec());
+
+    let bytes = Transaction::Phoenix(tx.clone()).to_var_bytes();
+
+    let len = bytes.len().to_le_bytes();
+
+    let ptr = mem::malloc(4 + bytes.len() as u32);
+    let ptr = ptr as *mut u8;
+
+    *proved_ptr = ptr;
+
+    ptr::copy_nonoverlapping(len.as_ptr(), ptr, 4);
+    ptr::copy_nonoverlapping(bytes.as_ptr(), ptr.add(4), bytes.len());
+
+    let displayed = revert(&tx.hash());
+    let bytes = displayed.as_bytes();
+
+    ptr::copy_nonoverlapping(bytes.as_ptr(), hash_ptr.as_mut_ptr(), 64);
+
+    ErrorCode::Ok
+}
+
+#[no_mangle]
+pub unsafe fn phoenix(
+    rng: &[u8; 32],
+    seed: &Seed,
+    sender_index: u8,
+    receiver: &[u8; PhoenixPublicKey::SIZE],
+    inputs: *const u8,
+    openings: *const u8,
+    root: &[u8; BlsScalar::SIZE],
+    transfer_value: *const u64,
+    obfuscated_transaction: bool, // ?
+    deposit: *const u64,          // ?
+    gas_limit: *const u64,
+    gas_price: *const u64,
+    chain_id: u8,    // ?
+    data: *const u8, // ?
+    tx_ptr: *mut *mut u8,
+    proof_ptr: *mut *mut u8,
+) -> ErrorCode {
+    let mut rng = ChaCha12Rng::from_seed(*rng);
+
+    let sender_sk = derive_phoenix_sk(&seed, sender_index);
+    let change_pk = PhoenixPublicKey::from(&sender_sk);
+    let receiver_pk = PhoenixPublicKey::from_bytes(receiver)
+        .or(Err(ErrorCode::DeserializationError))?;
+
+    let root: BlsScalar =
+        rkyv::from_bytes(root).or(Err(ErrorCode::UnarchivingError))?;
+
+    let openings: Vec<Option<NoteOpening>> = mem::from_buffer(openings)?;
+
+    let notes: Vec<NoteLeaf> = mem::from_buffer(inputs)?;
+
+    let inputs: Vec<(Note, NoteOpening)> = notes
+        .into_iter()
+        .map(|note_leaf| note_leaf.note)
+        .zip(openings.into_iter())
+        .filter_map(|(note, opening)| opening.map(|op| (note, op)))
+        .collect();
+
+    let data: Option<TransactionData> =
+        if data.is_null() { None } else { todo!() };
```

this might lead to `unreachable` in wasm code maybe throw an error here?